### PR TITLE
[CARBONDATA-1729]Fix the compatibility issue with hadoop <= 2.6 and 2.7

### PR DESCRIPTION
**Pull request by zzcclp**
1. Recover profile of 'hadoop-2.2.0' to pom.xml
2. Use reflection mechanism to implement 'truncate' method

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by zzcclp** [comment]
@QiangCai @jackylk please review, thanks.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1202/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1206/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1213/

**Comment by zzcclp** [comment]
I have complied successfully with below commands:
1. mvn -Pspark-2.1 -Pbuild-with-format -Dspark.version=2.1.2 clean package;
2. mvn -Pspark-2.1 -Phadoop-2.2.0 -Pbuild-with-format -Dspark.version=2.1.2 -Dhadoop.version=2.6.0-cdh5.7.1;

@QiangCai @jackylk @chenliang613 please review, thanks.

File: pom.xml [hunk 1]
```diff
@@ -453,9 +453,9 @@
       </build>
     </profile>
     <profile>
-      <id>hadoop-2.7.2</id>
+      <id>hadoop-2.2.0</id>
```
Comment (chenliang613) [inline pom.xml]: you can add a profile for hadoop-2.2.0, don't need to overwrite hadoop-2.7.2.  by default, should use hadoop-2.7.2

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java [hunk 2]
```diff
@@ -462,39 +461,8 @@ public static DataOutputStream getDataOutputStreamUsingAppend(String path, FileT
...
-          }
-        }
-        return;
-      case HDFS:
-      case ALLUXIO:
-      case VIEWFS:
-      case S3:
-        Path pt = new Path(path);
-        FileSystem fs = pt.getFileSystem(configuration);
-        fs.truncate(pt, newSize);
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: I think it is better to use java reflection for line 485 only, no need to modify previous file

File: core/src/main/java/org/apache/carbondata/core/datastore/filesystem/AbstractDFSCarbonFile.java [hunk 3]
```diff
@@ -154,52 +155,68 @@ public boolean delete() {
...
-      CarbonUtil.closeStreams(dataInputStream, dataOutputStream);
-      // rename the temp file to original file
-      tempFile.renameForce(fileName);
-      fileTruncatedSuccessfully = true;
     } catch (IOException e) {
-      LOGGER.error("Exception occurred while truncating the file " + e.getMessage());
-    } finally {
-      CarbonUtil.closeStreams(dataOutputStream, dataInputStream);
+      LOGGER.error("IOException occurred while truncating the file " + e.getMessage());
+    } catch (Exception e) {
```
Comment (chenliang613) [inline core/src/main/java/org/apache/carbondata/core/datastore/filesystem/AbstractDFSCarbonFile.java]: why finally, need add the catch again : 

catch (Exception e) {
	  LOGGER.error("Other exception occurred while truncating the file + e.getMessage());
      }

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java [hunk 4]
```diff
@@ -462,39 +461,8 @@ public static DataOutputStream getDataOutputStreamUsingAppend(String path, FileT
    * @throws IOException
    */
   public static void truncateFile(String path, FileType fileType, long newSize) throws IOException {
-    path = path.replace("\\", "/");
```
Comment (chenliang613) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: why remove these code.

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java (hunk 2 above)
Comment (zzcclp) [reply core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: According to discussion with @QiangCai offline, just use the interface 'CarbonFile.truncate' to truncate file uniformly.
@QiangCai what do you think about this?

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java (hunk 4 above)
Comment (zzcclp) [reply core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: want to use the interface 'CarbonFile.truncate' to truncate file uniformly.

File: core/src/main/java/org/apache/carbondata/core/datastore/filesystem/AbstractDFSCarbonFile.java (hunk 3 above)
Comment (zzcclp) [reply core/src/main/java/org/apache/carbondata/core/datastore/filesystem/AbstractDFSCarbonFile.java]: Method.invoke may throw other exceptions when running, such as: IllegalAccessException, 
 IllegalArgumentException, InvocationTargetException;
So need to add this catch for above exceptions.

The finally block is for the try block which is in catch NoSuchMethodException branch.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1296/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1298/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1300/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1765/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1303/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1305/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1767/

**Comment by zzcclp** [comment]
@jackylk @chenliang613  @QiangCai   According to Jacky's suggestion, just use java reflection for FileSystem.truncate in FileFactory.java,  please review, thanks.

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1770/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1775/

**Comment by QiangCai** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1778/

**Pull request closed without merging**